µfmt doesn't handle exceptions from black/µsort

```
(.venv) jreese@butterfree ~/workspace/torch/fairseq main  » ufmt check .
concurrent.futures.process._RemoteTraceback:
"""
Traceback (most recent call last):
  File "/Users/jreese/.pyenv/versions/3.9.9/lib/python3.9/concurrent/futures/process.py", line 243, in _process_worker
    r = call_item.fn(*call_item.args, **call_item.kwargs)
  File "/Users/jreese/.pyenv/versions/3.9.9/lib/python3.9/concurrent/futures/process.py", line 202, in _process_chunk
    return [fn(*args) for args in chunk]
  File "/Users/jreese/.pyenv/versions/3.9.9/lib/python3.9/concurrent/futures/process.py", line 202, in <listcomp>
    return [fn(*args) for args in chunk]
  File "/Users/jreese/workspace/torch/fairseq/.venv/lib/python3.9/site-packages/ufmt/core.py", line 88, in ufmt_file
    dst_contents = ufmt_string(path, src_contents, usort_config, black_config)
  File "/Users/jreese/workspace/torch/fairseq/.venv/lib/python3.9/site-packages/ufmt/core.py", line 50, in ufmt_string
    content = format_file_contents(content, fast=False, mode=mode)
  File "src/black/__init__.py", line 993, in format_file_contents
  File "src/black/__init__.py", line 970, in check_stability_and_equivalence
  File "src/black/__init__.py", line 1361, in assert_equivalent
AssertionError: INTERNAL ERROR: Black produced code that is not equivalent to the source.  Please report a bug on https://github.com/psf/black/issues.  This diff might be helpful: /var/folders/jg/hrl2p4ms1vjbn44gmmyl4p7r0000gn/T/blk_d04ehvsc.log
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/jreese/workspace/torch/fairseq/.venv/bin/ufmt", line 8, in <module>
    sys.exit(main())
  File "/Users/jreese/workspace/torch/fairseq/.venv/lib/python3.9/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/Users/jreese/workspace/torch/fairseq/.venv/lib/python3.9/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/Users/jreese/workspace/torch/fairseq/.venv/lib/python3.9/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/jreese/workspace/torch/fairseq/.venv/lib/python3.9/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/jreese/workspace/torch/fairseq/.venv/lib/python3.9/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/Users/jreese/workspace/torch/fairseq/.venv/lib/python3.9/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/jreese/workspace/torch/fairseq/.venv/lib/python3.9/site-packages/ufmt/cli.py", line 54, in check
    results = ufmt_paths(paths, dry_run=True)
  File "/Users/jreese/workspace/torch/fairseq/.venv/lib/python3.9/site-packages/ufmt/core.py", line 117, in ufmt_paths
    results = list(runner.run(all_paths, fn).values())
  File "/Users/jreese/workspace/torch/fairseq/.venv/lib/python3.9/site-packages/trailrunner/core.py", line 183, in run
    results = list(exe.map(func, paths))
  File "/Users/jreese/.pyenv/versions/3.9.9/lib/python3.9/concurrent/futures/process.py", line 559, in _chain_from_iterable_of_lists
    for element in iterable:
  File "/Users/jreese/.pyenv/versions/3.9.9/lib/python3.9/concurrent/futures/_base.py", line 608, in result_iterator
    yield fs.pop().result()
  File "/Users/jreese/.pyenv/versions/3.9.9/lib/python3.9/concurrent/futures/_base.py", line 445, in result
    return self.__get_result()
  File "/Users/jreese/.pyenv/versions/3.9.9/lib/python3.9/concurrent/futures/_base.py", line 390, in __get_result
    raise self._exception
AssertionError: INTERNAL ERROR: Black produced code that is not equivalent to the source.  Please report a bug on https://github.com/psf/black/issues.  This diff might be helpful: /var/folders/jg/hrl2p4ms1vjbn44gmmyl4p7r0000gn/T/blk_d04ehvsc.log
```

--- a/ufmt/cli.py
+++ b/ufmt/cli.py
@@ -4,7 +4,7 @@
 import logging
 import sys
 from pathlib import Path
-from typing import List
+from typing import List, Tuple
 
 import click
 from moreorless.click import echo_color_precomputed_diff
@@ -22,10 +22,18 @@ def init_logging(*, debug: bool = False) -> None:
     logging.getLogger("blib2to3").setLevel(logging.WARNING)
 
 
-def echo_results(results: List[Result], diff: bool = False) -> bool:
+def echo_results(results: List[Result], diff: bool = False) -> Tuple[bool, bool]:
+    error = False
     changed = False
 
     for result in results:
+        if result.error is not None:
+            msg = str(result.error)
+            lines = msg.splitlines()
+            msg = lines[0]
+            click.secho(f"Error formatting {result.path}: {msg}", fg="yellow")
+            error = True
+
         if result.changed:
             changed = True
             if result.written:
@@ -35,7 +43,7 @@ def echo_results(results: List[Result], diff: bool = False) -> bool:
             if diff and result.diff:
                 echo_color_precomputed_diff(result.diff)
 
-    return changed
+    return changed, error
 
 
 @click.group()
@@ -52,8 +60,8 @@ def check(ctx: click.Context, names: List[str]):
     """Check formatting of one or more paths"""
     paths = [Path(name) for name in names] if names else [Path(".")]
     results = ufmt_paths(paths, dry_run=True)
-    changed = echo_results(results)
-    if changed:
+    changed, error = echo_results(results)
+    if changed or error:
         ctx.exit(1)
 
 
@@ -64,15 +72,18 @@ def diff(ctx: click.Context, names: List[str]):
     """Generate diffs for any files that need formatting"""
     paths = [Path(name) for name in names] if names else [Path(".")]
     results = ufmt_paths(paths, dry_run=True, diff=True)
-    changed = echo_results(results, diff=True)
-    if changed:
+    changed, error = echo_results(results, diff=True)
+    if changed or error:
         ctx.exit(1)
 
 
 @main.command()
+@click.pass_context
 @click.argument("names", type=click.Path(exists=True), nargs=-1, metavar="[PATH] ...")
-def format(names: List[str]):
+def format(ctx: click.Context, names: List[str]):
     """Format one or more paths in place"""
     paths = [Path(name) for name in names] if names else [Path(".")]
     results = ufmt_paths(paths)
-    echo_results(results)
+    _, error = echo_results(results)
+    if error:
+        ctx.exit(1)

--- a/ufmt/core.py
+++ b/ufmt/core.py
@@ -64,6 +64,8 @@ def ufmt_bytes(
         best guess at file encodings. Otherwise, use :func:`tokenize.detect_encodings`.
     """
     result = usort(content, usort_config, path)
+    if result.error:
+        raise result.error
 
     if path.suffix == ".pyi":
         black_config = replace(black_config, is_pyi=True)
@@ -153,17 +155,21 @@ def ufmt_file(
 
     LOG.debug(f"Checking {path}")
 
-    src_contents, encoding, newline = read_file(path)
-    dst_contents = ufmt_bytes(
-        path,
-        src_contents,
-        encoding=encoding,
-        black_config=black_config,
-        usort_config=usort_config,
-        post_processor=post_processor,
-    )
-
     result = Result(path)
+
+    try:
+        src_contents, encoding, newline = read_file(path)
+        dst_contents = ufmt_bytes(
+            path,
+            src_contents,
+            encoding=encoding,
+            black_config=black_config,
+            usort_config=usort_config,
+            post_processor=post_processor,
+        )
+    except Exception as e:
+        result.error = e
+        return result
 
     if src_contents != dst_contents:
         result.changed = True
@@ -177,8 +183,11 @@ def ufmt_file(
 
         if not dry_run:
             LOG.debug(f"Formatted {path}")
-            write_file(path, dst_contents, newline=newline)
-            result.written = True
+            try:
+                write_file(path, dst_contents, newline=newline)
+                result.written = True
+            except Exception as e:
+                result.error = e
 
     return result
 

--- a/ufmt/core.py
+++ b/ufmt/core.py
@@ -45,6 +45,10 @@ def ufmt_bytes(
     represents a type stub (has a ``.pyi`` suffix), the black config object will be
     updated to set ``is_pyi = True``.
 
+    This function will not catch any errors during formatting, except for "everything is
+    fine" messages, like :exc:`black.NothingChanged`. All other errors must be handled
+    by the code calling this function; see :func:`ufmt_file` for example error handling.
+
     Optionally takes a post processor matching the :class:`PostProcessor` protocol.
     If given, the post processor will be called with the updated byte string content
     after it has been run through µsort and black. The return value of the post
@@ -139,6 +143,11 @@ def ufmt_file(
     changes to disk. Passing ``diff = True`` will generate a unified diff of changes
     on the :class:`Result` object.
 
+    Any errors that occur during formatting will be caught, and those exceptions will
+    be attached to the :attr:`Result.error` property of the result object. It is the
+    responsibility of code calling this function to check for errors in results and
+    handle or surface them appropriately.
+
     Optionally takes ``black_config_factory`` or ``usort_config_factory`` to override
     the default configuration detection for each respective tool. Factory functions
     must take a :class:`pathlib.Path` object and return a valid :class:`BlackConfig`
@@ -210,6 +219,10 @@ def ufmt_paths(
     performance and CPU utilization.
 
     Returns a list of :class:`Result` objects for each file formatted.
+    Any errors that occur during formatting will be caught, and those exceptions will
+    be attached to the :attr:`Result.error` property of the result object. It is the
+    responsibility of code calling this function to check for errors in results and
+    handle or surface them appropriately.
 
     See :func:`ufmt_file` for details on parameters, config factories,
     and post processors. All parameters are passed through to :func:`ufmt_file`.

--- a/ufmt/tests/cli.py
+++ b/ufmt/tests/cli.py
@@ -10,6 +10,7 @@ from unittest.mock import call, patch
 
 import trailrunner
 from click.testing import CliRunner
+from libcst import ParserSyntaxError
 
 from ufmt.cli import echo_results, main
 from ufmt.core import Result
@@ -119,6 +120,29 @@ class CliTest(TestCase):
             )
             self.assertEqual(1, result.exit_code)
 
+        with self.subTest("syntax error"):
+            ufmt_mock.reset_mock()
+            ufmt_mock.return_value = [
+                Result(Path("bar.py"), changed=False),
+                Result(
+                    Path("foo/frob.py"),
+                    error=ParserSyntaxError(
+                        "bad",
+                        lines=("", "", "", "foo bar fizzbuzz hello world"),
+                        raw_line=4,
+                        raw_column=15,
+                    ),
+                ),
+            ]
+            result = runner.invoke(main, ["check", "bar.py", "foo/frob.py"])
+            ufmt_mock.assert_called_with(
+                [Path("bar.py"), Path("foo/frob.py")], dry_run=True
+            )
+            self.assertRegex(
+                result.stdout, r"Error formatting foo/frob\.py: Syntax Error @ 4:16"
+            )
+            self.assertEqual(1, result.exit_code)
+
     @patch("ufmt.cli.ufmt_paths")
     def test_diff(self, ufmt_mock):
         runner = CliRunner()
@@ -161,6 +185,29 @@ class CliTest(TestCase):
             )
             self.assertEqual(1, result.exit_code)
 
+        with self.subTest("syntax error"):
+            ufmt_mock.reset_mock()
+            ufmt_mock.return_value = [
+                Result(Path("bar.py"), changed=False),
+                Result(
+                    Path("foo/frob.py"),
+                    error=ParserSyntaxError(
+                        "bad",
+                        lines=("", "", "", "foo bar fizzbuzz hello world"),
+                        raw_line=4,
+                        raw_column=15,
+                    ),
+                ),
+            ]
+            result = runner.invoke(main, ["diff", "bar.py", "foo/frob.py"])
+            ufmt_mock.assert_called_with(
+                [Path("bar.py"), Path("foo/frob.py")], dry_run=True, diff=True
+            )
+            self.assertRegex(
+                result.stdout, r"Error formatting foo/frob\.py: Syntax Error @ 4:16"
+            )
+            self.assertEqual(1, result.exit_code)
+
     @patch("ufmt.cli.ufmt_paths")
     def test_format(self, ufmt_mock):
         runner = CliRunner()
@@ -197,3 +244,24 @@ class CliTest(TestCase):
             result = runner.invoke(main, ["format", "bar.py", "foo/frob.py"])
             ufmt_mock.assert_called_with([Path("bar.py"), Path("foo/frob.py")])
             self.assertEqual(0, result.exit_code)
+
+        with self.subTest("syntax error"):
+            ufmt_mock.reset_mock()
+            ufmt_mock.return_value = [
+                Result(Path("bar.py"), changed=False),
+                Result(
+                    Path("foo/frob.py"),
+                    error=ParserSyntaxError(
+                        "bad",
+                        lines=("", "", "", "foo bar fizzbuzz hello world"),
+                        raw_line=4,
+                        raw_column=15,
+                    ),
+                ),
+            ]
+            result = runner.invoke(main, ["format", "bar.py", "foo/frob.py"])
+            ufmt_mock.assert_called_with([Path("bar.py"), Path("foo/frob.py")])
+            self.assertRegex(
+                result.stdout, r"Error formatting foo/frob\.py: Syntax Error @ 4:16"
+            )
+            self.assertEqual(1, result.exit_code)

--- a/ufmt/tests/cli.py
+++ b/ufmt/tests/cli.py
@@ -139,7 +139,7 @@ class CliTest(TestCase):
                 [Path("bar.py"), Path("foo/frob.py")], dry_run=True
             )
             self.assertRegex(
-                result.stdout, r"Error formatting foo/frob\.py: Syntax Error @ 4:16"
+                result.stdout, r"Error formatting .*frob\.py: Syntax Error @ 4:16"
             )
             self.assertEqual(1, result.exit_code)
 
@@ -204,7 +204,7 @@ class CliTest(TestCase):
                 [Path("bar.py"), Path("foo/frob.py")], dry_run=True, diff=True
             )
             self.assertRegex(
-                result.stdout, r"Error formatting foo/frob\.py: Syntax Error @ 4:16"
+                result.stdout, r"Error formatting .*frob\.py: Syntax Error @ 4:16"
             )
             self.assertEqual(1, result.exit_code)
 
@@ -262,6 +262,6 @@ class CliTest(TestCase):
             result = runner.invoke(main, ["format", "bar.py", "foo/frob.py"])
             ufmt_mock.assert_called_with([Path("bar.py"), Path("foo/frob.py")])
             self.assertRegex(
-                result.stdout, r"Error formatting foo/frob\.py: Syntax Error @ 4:16"
+                result.stdout, r"Error formatting .*frob\.py: Syntax Error @ 4:16"
             )
             self.assertEqual(1, result.exit_code)

--- a/ufmt/tests/core.py
+++ b/ufmt/tests/core.py
@@ -8,6 +8,7 @@ from unittest import TestCase
 from unittest.mock import call, Mock, patch
 
 import trailrunner
+from libcst import ParserSyntaxError
 
 import ufmt
 from ufmt.types import BlackConfig, Encoding, UsortConfig
@@ -88,6 +89,10 @@ class Bar:
     def hello(
         self,
     ) -> None: ...
+"""
+
+INVALID_SYNTAX = """\
+print "hello world"
 """
 
 
@@ -225,6 +230,27 @@ class CoreTest(TestCase):
                 result = ufmt.ufmt_file(f)
                 self.assertFalse(result.changed)
                 self.assertEqual(windows_content, f.read_bytes())
+
+            with self.subTest("invalid syntax"):
+                f.write_text(INVALID_SYNTAX)
+
+                result = ufmt.ufmt_file(f)
+                self.assertIsInstance(result.error, ParserSyntaxError)
+
+            with self.subTest("file not found"):
+                f.unlink()
+
+                result = ufmt.ufmt_file(f)
+                self.assertIsInstance(result.error, FileNotFoundError)
+
+            with self.subTest("write failed"):
+                with patch("ufmt.core.write_file") as write_mock:
+                    write_mock.side_effect = PermissionError("fake permission error")
+
+                    f.write_text(POORLY_FORMATTED_CODE)
+
+                    result = ufmt.ufmt_file(f)
+                    self.assertIsInstance(result.error, PermissionError)
 
     def test_ufmt_paths(self):
         with TemporaryDirectory() as td:

--- a/ufmt/types.py
+++ b/ufmt/types.py
@@ -41,3 +41,4 @@ class Result:
     changed: bool = False
     written: bool = False
     diff: Optional[str] = None
+    error: Optional[Exception] = None

